pkg/costattribution/tracker.go

```diff
@@ -0,0 +1,200 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package costattribution
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/model/labels"
+	"go.uber.org/atomic"
+)
+
+type observation struct {
+	lvalues    []string
+	lastUpdate *atomic.Int64
+}
+
+func (t *Tracker) cleanupTrackerAttribution(vals []string) {
+	t.activeSeriesPerUserAttribution.DeleteLabelValues(vals...)
+	t.receivedSamplesAttribution.DeleteLabelValues(vals...)
+	t.discardedSampleAttribution.DeleteLabelValues(vals...)
+}
+
+func (t *Tracker) cleanupTracker(userID string) {
+	filter := prometheus.Labels{"user": userID}
+	t.activeSeriesPerUserAttribution.DeletePartialMatch(filter)
+	t.receivedSamplesAttribution.DeletePartialMatch(filter)
+	t.discardedSampleAttribution.DeletePartialMatch(filter)
+}
+
+type Tracker struct {
+	userID                         string
+	trackedLabels                  []string
+	maxCardinality                 int
+	activeSeriesPerUserAttribution *prometheus.GaugeVec
+	receivedSamplesAttribution     *prometheus.CounterVec
+	discardedSampleAttribution     *prometheus.CounterVec
+
+	// oLock protects the observed map
+	oLock    sync.RWMutex
```

Comment:
We usually call mutexes as "mutex", not "lock".
Also, opionionated, but I think it could have the entire name `obseveredMtx` instead of just `o`: this isn't used so often.

Reply:
addressed here https://github.com/grafana/mimir/pull/9733/commits/f4ebc07871ef79a0ed7ee31c48032a9fd42fe757#diff-6b257cdf260797a2205e2032db7c6f13c129bff2a6e6eb6b54f6d68ee02c3c48L42